st2common/st2common/services/action.py

```diff
@@ -97,11 +102,110 @@ def schedule(liveaction):
+                      liveaction.id)
+        raise
+
+    # stamp liveaction with process_info
+    runner_info = system_info.get_process_info()
+
+    # Update liveaction status to "scheduled"
+    liveaction_db = action_utils.update_liveaction_status(
+        status=action_constants.LIVEACTION_STATUS_SCHEDULED,
+        runner_info=runner_info,
+        liveaction_id=liveaction_db.id)
+
+    action_execution_db = executions.update_execution(liveaction_db)
+
+    extra = {'action_execution_db': action_execution_db, 'liveaction_db': liveaction_db}
+    LOG.audit('Scheduled action execution.', extra=extra)
+
+    # the extra field will not be shown in non-audit logs so temporarily log at info.
+    LOG.info('Scheduled {~}action_execution: %s / {~}live_action: %s with "%s" status.',
+             action_execution_db.id, liveaction_db.id, liveaction.status)
+
+    LiveAction.publish_schedule(liveaction_db)
+
+
+def execute(liveaction, container):
+    """
+    Execute an action.
+
+    :return: result
+    :rtype: dict
+    """
+    # Only execution actions which haven't completed yet.
+    if liveaction.status == action_constants.LIVEACTION_STATUS_CANCELED:
+        LOG.info('Not executing liveaction %s. User canceled execution.', liveaction.id)
+        if not liveaction.result:
+            action_utils.update_liveaction_status(
+                status=action_constants.LIVEACTION_STATUS_CANCELED,
+                result={'message': 'Action execution canceled by user.'},
+                liveaction_id=liveaction.id)
+        return
+
+    if liveaction.status != action_constants.LIVEACTION_STATUS_SCHEDULED:
+        LOG.info('Action dispatching service is ignoring liveaction %s with "%s" status.',
+                 liveaction.id, liveaction.status)
+        return
+
+    try:
+        liveaction_db = action_utils.get_liveaction_by_id(liveaction.id)
+    except StackStormDBObjectNotFoundError:
+        LOG.exception('Failed to find liveaction %s in the database.',
+                      liveaction.id)
+        raise
+
+    # stamp liveaction with process_info
+    runner_info = system_info.get_process_info()
+
+    # Update liveaction status to "running"
+    liveaction_db = action_utils.update_liveaction_status(
+        status=action_constants.LIVEACTION_STATUS_RUNNING,
+        runner_info=runner_info,
+        liveaction_id=liveaction_db.id)
+
+    action_execution_db = executions.update_execution(liveaction_db)
+
+    # Launch action
+    extra = {'action_execution_db': action_execution_db, 'liveaction_db': liveaction_db}
+    LOG.audit('Launching action execution.', extra=extra)
+
+    # the extra field will not be shown in non-audit logs so temporarily log at info.
+    LOG.info('Dispatched {~}action_execution: %s / {~}live_action: %s with "%s" status.',
+             action_execution_db.id, liveaction_db.id, liveaction.status)
+    try:
+        result = container.dispatch(liveaction_db)
```

Comment:
container.dispatch is very st2actions specific. I am not sure how/why you would want to encourage reuse.